### ISSUE TITLE
How to delete an object from S3 bucket with Amplify?

### ISSUE THREAD

**Issue by chai86** [aws-amplify/amplify-js#4578]
I have a mobile app which has users authenticate via AWS Cognito and they end up in a user pool (not Identity pool). They're able to put objects into the bucket no problem, but they cant delete. What i want to do is for each logged in user to be able to delete an file in the bucket. 

The path to the files is for example: `my_bucket_name/protected/videos/video1`

To delete, i call this: 

    import { Storage } from 'aws-amplify'; 
.
.
.

    delFromS3 = async () => {
          Storage.remove('protected/videos/video1')  
             .then(result => console.log('Deleted Video from S3'))
             .catch(err => console.log('Deleting video from S3 error: ', err));
      }
    
I kept receiving the error Access Denied when calling this., so i added a Bucket Policy: 

    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Principal": {
                    "Federated": "cognito-identity.amazonaws.com"
                },
                "Action": "s3:DeleteObject",
                "Resource": "arn:aws:s3:::my_bucket_name/*"
            }
        ]
    }

The only time this function returns a result is when i put `"Principal": "*"` but this makes my bucket PUBLIC to anyone, which i don't want to do. It also doesn't accept this as a valid Prinipal policy:

    "Principal": { "AWS": [ "arn:awseu-west-2_2ecGAT74q" ] }. 

So i need to know what the correct Principal is. 

So i either need a way to allow users in the cognito user pool to be authorised to delete an object. Or because i manually know the path to the file in each users sub-bucket (e.g. `/protected/videos/video1`) just pass that in my `delFromS3()` function. Is my bucket policy wrong?

OR do i need to use an Identity pool? How do I allow all the users in my existing user pool to have an Identity to enable them to delete in S3? At the moment users use Cognito to get into the user pool using Auth.signIn()

**Comment by elorzafe** [comment]
@chai86 you dont need to specify the level (in this case `protected`) and the identity id. Those are added by the library automatically. more info [here](https://aws-amplify.github.io/docs/js/storage#put)

**Comment by chai86** [comment]
@elorzafe so when i'm calling the remove() I tried this:
`Storage.remove('videos/video1')`
but still no luck. The user is already signed in via cognito in the mobile app. 

The full path is actually `my_bucket_name/protected/eu-west-2:de55c2rf-8f1e-836d-88f9-82da662aau6dt/videos/video1`

The `eu-west-2:de55c2rf-8f1e-836d-88f9-82da662aau6dt` folder gets automatically created for each user when i do `Storage.Put(videos/video1)`

Any ideas?

**Comment by elorzafe** [comment]
@chai86  

What is the CORS configuration of the bucket. Can you try the policy mentioned [here](https://aws-amplify.github.io/docs/js/storage#amazon-s3-bucket-cors-policy-setup), I think it could be that the problem.

**Comment by chai86** [comment]
@elorzafe CORS configuration of the bucket is as follows:

```
<?xml version="1.0" encoding="UTF-8"?>
<CORSConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
<CORSRule>
    <ID>S3CORSRuleId1</ID>
    <AllowedOrigin>*</AllowedOrigin>
    <AllowedMethod>GET</AllowedMethod>
    <AllowedMethod>HEAD</AllowedMethod>
    <AllowedMethod>PUT</AllowedMethod>
    <AllowedMethod>POST</AllowedMethod>
    <AllowedMethod>DELETE</AllowedMethod>
    <MaxAgeSeconds>3000</MaxAgeSeconds>
    <ExposeHeader>x-amz-server-side-encryption</ExposeHeader>
    <ExposeHeader>x-amz-request-id</ExposeHeader>
    <ExposeHeader>x-amz-id-2</ExposeHeader>
    <ExposeHeader>ETag</ExposeHeader>
    <AllowedHeader>*</AllowedHeader>
</CORSRule>
</CORSConfiguration>
```

**Comment by elorzafe** [comment]
@chai86 

Can you try setting the level on `Storage.remove` as well. Maybe the default level you have configured on Storage is not `protected`.

```javascript
Storage.remove('videos/video1', { level: 'protected' });
```

**Comment by chai86** [comment]
I tried that already (because of how i do the Storage.Put), still AccessDenied

**Comment by elorzafe** [comment]
@chai86 

> OR do i need to use an Identity pool? How do I allow all the users in my existing user pool to have an Identity to enable them to delete in S3? At the moment users use Cognito to get into the user pool using Auth.signIn()

Yes, you will need an Identity pool to get AWS Credentials. (I am not sure how your able to upload things to S3 though). The Cognito User Pool is an Authentication provider on the Identity Pool, when you have that configured Amplify gets credentials automatically.

That identity pool must have an auth role that has attached the policies mentioned [here](https://aws-amplify.github.io/docs/js/storage#using-amazon-s3) 

Is much simpler to configure those using [amplify-cli](https://aws-amplify.github.io/docs/)

**Comment by chai86** [comment]
@elorzafe I do think thats the piece missing. I've been trying to implement this for many days now. I tried the bucket policy that you linked a few days back but it said it's missing a "Prinicipal" field. I tried a few: 

```
"Principal": { "AWS": [ "arn:aws:cognito-idp:......." ] },

"Principal": {  "Federated": "cognito-identity.amazonaws.com"   },
```
neither worked.

So i do have a user pool and an associated Identity pool ( I know this because at the moment each has 19 users/identities in them).  What steps must i follow to get an auth role? 

Also, for all the other users and new users that will join the pool in the future, will they automatically  be given the rights to Storage.remove()?
